internal/controller/oauth/oidc_provider.go

```diff
@@ -0,0 +1,106 @@
+// Copyright Envoy AI Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package oauth
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	egv1a1 "github.com/envoyproxy/gateway/api/v1alpha1"
+	"golang.org/x/oauth2"
+)
+
+// OIDCProvider extends ClientCredentialsTokenProvider with OIDC support.
+type OIDCProvider struct {
+	tokenProvider  *ClientCredentialsTokenProvider
+	oidcCredential *egv1a1.OIDC
+}
+
+// NewOIDCProvider creates a new OIDC-aware provider.
+func NewOIDCProvider(tokenProvider *ClientCredentialsTokenProvider, oidcCredentials *egv1a1.OIDC) *OIDCProvider {
+	return &OIDCProvider{
+		tokenProvider:  tokenProvider,
+		oidcCredential: oidcCredentials,
+	}
+}
+
+// getOIDCProviderConfig retrieves or creates OIDC config for the given issuer URL.
+func (p *OIDCProvider) getOIDCProviderConfig(ctx context.Context, issuerURL string) (*oidc.ProviderConfig, []string, error) {
+	// Check context before proceeding in case context is cancelled because of timeout.
+	if err := ctx.Err(); err != nil {
+		return nil, nil, fmt.Errorf("context error before discovery: %w", err)
+	}
```

Comment:
you don't need this unless this is a part of the async operation that you are doing by yourself here. Operations will be canceled by for example http client and errors like timeout will be returned from there.